src/linear_algebra/char_poly/coeff.lean

```diff
@@ -160,3 +162,35 @@ begin
   rw [coeff_zero_eq_eval_zero, char_poly, eval_det, mat_poly_equiv_char_matrix, ← det_smul],
   simp
 end
+
+variables {p : ℕ} [fact p.prime]
+
+lemma zmod.char_poly_pow_card_of_nonempty [nonempty n] (M : matrix n n (zmod p)) :
+  char_poly (M ^ p) = char_poly M :=
+begin
+  apply frobenius_inj (polynomial (zmod p)) p, repeat {rw frobenius_def},
+  rw ← zmod.expand_p,
+  unfold char_poly, rw [alg_hom.map_det, ← det_pow],
+  apply congr_arg det,
+  apply mat_poly_equiv.injective, swap, { apply_instance },
+  rw [← mat_poly_equiv.coe_alg_hom, alg_hom.map_pow, mat_poly_equiv.coe_alg_hom,
+        mat_poly_equiv_char_matrix, sub_pow_char_of_commute, ← C_pow],
+  swap, { apply polynomial.commute_X },
+  -- the following is a nasty case bash that should be abstracted as a lemma
+  -- (and maybe it can be proven more... algebraically?)
+  ext, rw [coeff_sub, coeff_C],
+  by_cases hij : i = j; simp [char_matrix, hij, coeff_X_pow]; simp only [coeff_C]; split_ifs; simp *,
+end
+
+lemma zmod.char_poly_pow_card (M : matrix n n (zmod p)) :
+  char_poly (M ^ p) = char_poly M :=
```

Comment:
Why bother stating it separately for nonempty matrices then, when this proof is so short?